### PR TITLE
Simplify walker to use PatternMatcher instead of custom matching

### PR DESCRIPTION
**Pull request by aaronlehmann**
This changes the pattern matching implementation in walker to use `PatternMatcher`, making it consistent with the "copy" code.

This makes it support `**` patterns as well, which formerly didn't work.

The added benchmark shows the new implementation is slightly faster than the old one.

Opened on top of #107 - the first commit is from that PR, so only look at the second one.

Replaces #79

cc @coryb

### CODE REVIEW

**Comment by aaronlehmann** [comment]
Rebased now that #107 is merged.

File: walker_test.go [hunk 1]
```diff
@@ -305,3 +304,361 @@ func tmpDir(inp []*change) (dir string, retErr error) {
...
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		dir bar
+		file bar/foo
+	`, string(b.Bytes()))
+
+	// TODO can we support this? Currently filepath.Clean is called
```
Comment (coryb) [inline walker_test.go:568]: I think we can replace this comment/block with:
```
	b.Reset()
	err = Walk(context.Background(), d, &WalkOpt{
		IncludePatterns: []string{"**/foo/**"},
	}, bufWalk(b))
	assert.NoError(t, err)

	trimEqual(t, `
		dir foo
		dir foo/bar
		file foo/bar/bee
	`, string(b.Bytes()))
```

I had originally thought `**/foo/` could match all directories named `foo`, but I think `**/foo/**` better represents this idea and seems to work.

**Review by coryb — APPROVED** [review]
Looks good to me, thanks for carrying this work!

File: walker_test.go (hunk 1 above)
Comment (aaronlehmann) [reply walker_test.go]: Updated, thanks.

**Comment by aaronlehmann** [comment]
@tonistiigi: Do you see any compatibility concerns with this change?

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**